### PR TITLE
Fast loop task table

### PR DESCRIPTION
**Pull request by andyp1per**
This moves the various steps in the fast_loops into a task table that then appears in @SYS/tasks.txt
This allows more detailed timing information of the fast_loop to be determined.
```
STABILIZE> Getting @SYS/tasks.txt as -
TasksV2
AP_InertialSensor::update*       MIN=   6 MAX=  15 AVG=   7 OVR=  0 SLP=  0, TOT= 2.1%
Copter::run_rate_controller*     MIN=   9 MAX=  14 AVG=   9 OVR=  0 SLP=  0, TOT= 2.9%
Copter::motors_output*           MIN=  27 MAX=  38 AVG=  29 OVR=  0 SLP=  0, TOT= 8.4%
Copter::read_AHRS*               MIN= 110 MAX= 439 AVG= 151 OVR=160 SLP=  0, TOT=44.0%
Copter::read_inertia*            MIN=   6 MAX=  20 AVG=   6 OVR=  0 SLP=  0, TOT= 2.0%
Copter::check_ekf_reset*         MIN=   3 MAX=  16 AVG=   3 OVR=  0 SLP=  0, TOT= 0.9%
Copter::update_flight_mode*      MIN=  31 MAX=  66 AVG=  34 OVR=  0 SLP=  0, TOT=10.1%
Copter::update_home_from_EKF*    MIN=   2 MAX=  21 AVG=   3 OVR=  0 SLP=  0, TOT= 1.0%
Copter::update_land_and_crash_de MIN=   3 MAX=  15 AVG=   3 OVR=  0 SLP=  0, TOT= 0.9%
AP_Mount::update_fast*           MIN=   1 MAX=  11 AVG=   0 OVR=  0 SLP=  0, TOT= 0.2%
Copter::Log_Sensor_Health*       MIN=   2 MAX=  28 AVG=   2 OVR=  0 SLP=  0, TOT= 0.8%
Copter::Log_Video_Stabilisation* MIN=   1 MAX=  26 AVG=   1 OVR=  0 SLP=  0, TOT= 0.3%
AP_GyroFFT::sample_gyros*        MIN=   1 MAX=  21 AVG=   0 OVR=  0 SLP=  0, TOT= 0.2%
Copter::rc_loop                  MIN=   6 MAX=  32 AVG=   7 OVR=  0 SLP=  0, TOT= 0.5%
Copter::throttle_loop            MIN=   5 MAX=  35 AVG=   9 OVR=  0 SLP=  0, TOT= 0.3%
AP_GPS::update                   MIN=   6 MAX=  40 AVG=  16 OVR=  0 SLP=  0, TOT= 0.6%
OpticalFlow::update              MIN=   1 MAX=  28 AVG=   2 OVR=  0 SLP=  0, TOT= 0.3%
```

### CODE REVIEW

File: ArduCopter/Copter.cpp [hunk 1]
```diff
@@ -226,73 +229,55 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
...
-    motors_output();
-
-    // run EKF state estimator (expensive)
-    // --------------------
-    read_AHRS();
-
+const AP_Scheduler::FastTask Copter::fast_tasks[] = {
+    FAST_TASK_CLASS(AP_InertialSensor, &copter.ins, update),
+    FAST_TASK(run_rate_controller),
+    FAST_TASK(motors_output),
```
Comment (amilcarlucas) [inline ArduCopter/Copter.cpp:235]:
```suggestion
    // send outputs to the motors library immediately
    FAST_TASK(motors_output),
```

File: ArduCopter/Copter.cpp [hunk 2]
```diff
@@ -226,73 +229,55 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
...
-    // send outputs to the motors library immediately
-    motors_output();
-
-    // run EKF state estimator (expensive)
-    // --------------------
-    read_AHRS();
-
+const AP_Scheduler::FastTask Copter::fast_tasks[] = {
+    FAST_TASK_CLASS(AP_InertialSensor, &copter.ins, update),
+    FAST_TASK(run_rate_controller),
```
Comment (amilcarlucas) [inline ArduCopter/Copter.cpp:234]:
```suggestion
    // run low level rate controllers that only require IMU data
    FAST_TASK(run_rate_controller),
```

File: ArduCopter/Copter.cpp [hunk 3]
```diff
@@ -226,73 +229,55 @@ const AP_Scheduler::Task Copter::scheduler_tasks[] = {
...
-
-    // send outputs to the motors library immediately
-    motors_output();
-
-    // run EKF state estimator (expensive)
-    // --------------------
-    read_AHRS();
-
+const AP_Scheduler::FastTask Copter::fast_tasks[] = {
+    FAST_TASK_CLASS(AP_InertialSensor, &copter.ins, update),
```
Comment (amilcarlucas) [inline ArduCopter/Copter.cpp:233]:
```suggestion
    // update INS immediately to get current gyro data populated
    FAST_TASK_CLASS(AP_InertialSensor, &copter.ins, update),
```

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

**Comment by amilcarlucas** [comment]
Small size increase for the added functionality IMHO:
```
Binary Name      Text [B]        Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  --------------  -----------  -------------  ----------------------------  -------------------------
antennatracker   396 (+0.0294%)  0 (0.0000%)  4 (+0.0015%)   396 (+0.0293%)                                   615088
arducopter       828 (+0.0470%)  0 (0.0000%)  -4 (-0.0015%)  828 (+0.0469%)                                   201180
ardusub          416 (+0.0268%)  0 (0.0000%)  0 (0.0000%)    416 (+0.0268%)                                   412528
arducopter-heli  836 (+0.0473%)  0 (0.0000%)  -4 (-0.0015%)  836 (+0.0472%)                                   192908
ardurover        448 (+0.0282%)  0 (0.0000%)  0 (0.0000%)    448 (+0.0282%)                                   374960
blimp            400 (+0.0310%)  0 (0.0000%)  0 (0.0000%)    400 (+0.0310%)                                   673688
arduplane        604 (+0.0347%)  0 (0.0000%)  -4 (-0.0015%)  604 (+0.0346%)                                   219992
```

**Comment by andyp1per** [comment]
> Small size increase for the added functionality IMHO:

Yeah most of this will be the strings and the table. The size difference will be small on 1Mb boards. Could always use the 1Mb naming scheme to save a little - I'm not sure how much the fully qualified names add as most of them are Copter::

**Comment by tridge** [comment]
better to put them in the normal task table, with a special low priority

**Comment by rmackay9** [comment]
I think all PRs should have some sort of description..

**Review by tridge — CHANGES_REQUESTED** [review]
nice approach, needs a 2nd pass thru

**Comment by tridge** [comment]
please test with valgrind

**Comment by tridge** [comment]
why the overruns in read_AHRS ?

File: ArduSub/ArduSub.cpp [hunk 4]
```diff
@@ -46,6 +46,31 @@ SCHED_TASK_CLASS arguments:
  */
 
 const AP_Scheduler::Task Sub::scheduler_tasks[] = {
+    // update INS immediately to get current gyro data populated
+    FAST_TASK_CLASS(AP_InertialSensor, &sub.ins, update),
```
Comment (tridge) [inline ArduSub/ArduSub.cpp:50]: better to just call fast_loop fn so simpler change for sub and blimp

**Comment by andyp1per** [comment]
valgrind clean
no overruns - I verified

**Comment by andyp1per** [comment]
> why the overruns in read_AHRS ?

No overruns when I retested

File: libraries/AP_Scheduler/AP_Scheduler.h [hunk 5]
```diff
@@ -193,6 +207,10 @@ class AP_Scheduler
     const struct Task *_vehicle_tasks;
     uint8_t _num_vehicle_tasks;
 
+    // list of fast tasks to run
+    const struct FastTask *_fast_tasks;
+    uint8_t _num_fast_tasks;
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_Scheduler/AP_Scheduler.h]: removed

File: libraries/AP_Scheduler/AP_Scheduler.cpp [hunk 6]
```diff
@@ -422,7 +415,7 @@ void AP_Scheduler::update_logging()
     if (!(_options & uint8_t(Options::RECORD_TASK_INFO)) && perf_info.has_task_info()) {
         perf_info.free_task_info();
     } else if ((_options & uint8_t(Options::RECORD_TASK_INFO)) && !perf_info.has_task_info()) {
-        perf_info.allocate_task_info(_num_tasks);
+        perf_info.allocate_task_info(_num_tasks + _num_fast_tasks);
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_Scheduler/AP_Scheduler.cpp]: removed

**Review by MichelleRos — APPROVED** [review]

File: libraries/AP_Scheduler/AP_Scheduler.h [hunk 7]
```diff
@@ -87,6 +97,13 @@ class AP_Scheduler
         RECORD_TASK_INFO = 1 << 0
     };
 
+    enum FastTaskPriorities {
+        FAST_TASK_PRI0 = 0,
+        FAST_TASK_PRI1 = 1,
+        FAST_TASK_PRI2 = 2,
+        MAX_FAST_TASK_PRIORITIES = 3
+    };
+
```
Comment (peterbarker) [inline libraries/AP_Scheduler/AP_Scheduler.h:106]: Not used....

**Review by peterbarker — CHANGES_REQUESTED** [review]
Looks pretty good.

File: ArduCopter/Copter.h [hunk 8]
```diff
@@ -677,6 +675,7 @@ class Copter : public AP_Vehicle {
     void set_accel_throttle_I_from_pilot_throttle();
     void rotate_body_frame_to_NE(float &x, float &y);
     uint16_t get_pilot_speed_dn() const;
+    void run_rate_controller() { attitude_control->rate_controller_run(); }
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply ArduCopter/Copter.h:681]: I couldn't make it compile. I tried a lot of things, none worked. The pointer is set late so you have to have some kind of double dereference

File: libraries/AP_Scheduler/AP_Scheduler.h (hunk 7 above)
Comment (andyp1per) [reply libraries/AP_Scheduler/AP_Scheduler.h:106]: Used in the macro and to check the priorities

File: ArduSub/ArduSub.cpp (hunk 4 above)
Comment (andyp1per) [reply ArduSub/ArduSub.cpp:50]: Yep

File: ArduCopter/Copter.h [hunk 9]
```diff
@@ -637,10 +637,8 @@ class Copter : public AP_Vehicle {
     void update_using_interlock();
 
     // Copter.cpp
-    void get_scheduler_tasks(const AP_Scheduler::Task *&tasks,
-                             uint8_t &task_count,
-                             uint32_t &log_bit) override;
-    void fast_loop() override;
+    void get_scheduler_tasks(const AP_Scheduler::Task *&tasks, uint8_t &task_count, uint32_t &log_bit) override;
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply ArduCopter/Copter.h]: Fixed

**Comment by amilcarlucas** [comment]
It is smaller, and nicer :)
```
Binary Name      Text [B]        Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  --------------  -----------  -------------  ----------------------------  -------------------------
ardusub          612 (+0.0395%)  0 (0.0000%)  -4 (-0.0015%)  612 (+0.0394%)                                   412620
arducopter       552 (+0.0313%)  0 (0.0000%)  0 (0.0000%)    552 (+0.0313%)                                   199732
blimp            332 (+0.0257%)  0 (0.0000%)  4 (+0.0015%)   332 (+0.0257%)                                   673392
ardurover        -48 (-0.0030%)  0 (0.0000%)  0 (0.0000%)    -48 (-0.0030%)                                   374776
arducopter-heli  604 (+0.0341%)  0 (0.0000%)  4 (+0.0015%)   604 (+0.0341%)                                   192212
arduplane        72 (+0.0041%)   0 (0.0000%)  0 (0.0000%)    72 (+0.0041%)                                    218744
antennatracker   24 (+0.0018%)   0 (0.0000%)  0 (0.0000%)    24 (+0.0018%)                                    614992
```

**Review by peterbarker — APPROVED** [review]
LGTM

**Comment by andyp1per** [comment]
> why the overruns in read_AHRS ?

Actually there are none @tridge - must have been something I messed up earlier in the PR

**Review by amilcarlucas — APPROVED** [review]

**Comment by andyp1per** [comment]
Flight tested ok on copter minquad

**Comment by andyp1per** [comment]
Copter wants to leave the task table as-is in the PR.

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
@rmackay9 wants a bit more of a look

**Pull request merged**